README.md
# proxy-verifier
Verify that proxies are working and check their anonymity level, country, etc.


changelog.md
# Changelog

* v0.5.0:
  * Use ifconfig.me as default proxy check service
* v0.4.4:
  * Upgraded dependencies
* v0.4.3:
  * Upgraded dependencies
* v0.4.2:
  * Upgraded dependencies
* v0.4.1:
  * Upgraded dependencies
* v0.4.0:
  * `test()` method to allow for custom tests.
  * Deprecated `lookupCountry()`, `loadCountryData()`, and `loadCountryDataSync()`. These will be removed in a future release. Use [geoip-native-lite](https://github.com/chill117/geoip-native-lite) if you need to check the geo-location for your proxies.
* v0.3.0:
  * Performance improvements for `testAll()`.
  * Changed `proxy.ip_address` to `proxy.ipAddress`
  * Added `normalizeProxy(proxy)` method.
* v0.2.0:
  * Renamed `all` to `testAll`. `all` method now prints a deprecated warning.
  * Renamed `protocol` to `testProtocol`. `protocol` method now prints a deprecated warning.
  * Renamed `protocols` to `testProtocols`. `protocols` method now prints a deprecated warning.
  * Renamed `anonymityLevel` to `testAnonymityLevel`. `anonymityLevel` method now prints a deprecated warning.
  * Renamed `tunnel` to `testTunnel`. `tunnel` method now prints a deprecated warning.
  * Renamed `country` to `lookupCountry`. `country` method now prints a deprecated warning.


test/helpers.js
'use strict';

var _ = require('underscore');
var express = require('express');
var http = require('http');
var https = require('https');
var httpProxy = require('http-proxy');
var net = require('net');
var url = require('url');

var ssl = require('./ssl');

module.exports = {

	createAppServer: function(port, host) {

		var app = express();

		app.get('/check', function(req, res) {
			res.status(200).json(_.extend({}, { ip_addr: req.ip }, req.headers));
		});

		app.http = http.createServer(app).listen(port, host);
		app.https = https.createServer({ cert: ssl.cert, key: ssl.key }, app).listen(port + 1, host);

		return app;
	},

	createProxyServer: function(port, host, options) {

		options || (options = {});
		options = _.defaults(options, { tunnel: true });
		options.localAddress = host;

		if (options.proxyAuth) {
			options.auth = options.proxyAuth;
		}

		var proxy = httpProxy.createProxyServer(_.omit(options, 'tunnel'));

		proxy.http = http.createServer(function(req, res) {

			if (options.proxyAuth && !reqIsAuthenticated(req, options.proxyAuth)) {
				res.writeHead(407);
				return res.end();
			}

			var target = _.omit(url.parse(req.url), 'path');

			proxy.web(req, res, {
				target: target
			});

		}).listen(port, host);

		proxy.https = https.createServer({
			cert: ssl.cert,
			key: ssl.key
		}, function(req, res) {

			if (options.proxyAuth && !reqIsAuthenticated(req, options.proxyAuth)) {
				res.writeHead(407);
				return res.end();
			}

			var target = _.omit(url.parse(req.url), 'path');

			target.https = true;
			target.rejectUnauthorized = false;

			proxy.web(req, res, {
				target: target
			});

		}).listen(port + 1, host);

		if (options.tunnel) {
			var onConnect = _.bind(setupTunnel, undefined, host);
			proxy.http.on('connect', onConnect);
			proxy.https.on('connect', onConnect);
		} else {
			proxy.http.on('connect', methodNotAllowed);
			proxy.https.on('connect', methodNotAllowed);
		}

		return proxy;
	}
};

function methodNotAllowed(req, client, head) {
	client.write('HTTP/' + req.httpVersion + ' 405 Method Not Allowed\r\n\r\n');
	client.end();
}

function setupTunnel(localAddress, req, client, head) {

	var reqUri = url.parse(req.url.indexOf('://') !== -1 ? req.url : 'http://' + req.url);

	var server = net.connect({
		host: reqUri.hostname,
		port: reqUri.port,
		localAddress: localAddress
	}, function() {
		client.write(
			'HTTP/1.1 200 Connection Established\r\n' +
			'Proxy-agent: Node.js-Proxy\r\n' +
			'\r\n'
		);
		server.write(head);
		server.pipe(client);
		client.pipe(server);
	});
}

function reqIsAuthenticated(req, auth) {

	var authHeader = req.headers['proxy-authorization'] || req.headers['authorization'] || null;

	if (!authHeader) {
		return false;
	}

	var parts = Buffer.from(authHeader.substr('Basic '.length), 'base64').toString().split(':');
	return parts[0] === auth.user && parts[1] === auth.pass;
}


test/unit/all.js
'use strict';

var _ = require('underscore');
var expect = require('chai').expect;

var ProxyVerifier = require('../../');
var helpers = require('../helpers');

describe('testAll(proxy[, options], cb)', function() {

	var appServer;

	before(function() {
		appServer = helpers.createAppServer(3001, '127.0.0.1');
	});

	after(function() {
		appServer.http.close();
		appServer.https.close();
	});

	it('should be a function', function() {
		expect(ProxyVerifier.testAll).to.be.a('function');
	});

	describe('local proxy', function() {

		var proxyProtocol = 'http';
		var proxyServer;

		before(function() {
			proxyServer = helpers.createProxyServer(5050, '0.0.0.0');
		});

		after(function() {
			proxyServer.close();
			proxyServer.http.close();
			proxyServer.https.close();
		});

		it('default proxy check service', function(done) {

			var proxy = {
				ipAddress: proxyServer[proxyProtocol].address().address,
				port: proxyServer[proxyProtocol].address().port,
				protocols: [proxyProtocol]
			};

			var options = {};

			ProxyVerifier.testAll(proxy, options, function(error, result) {

				try {
					expect(error).to.equal(null);
					expect(result).to.be.an('object');
					expect(result.anonymityLevel).to.equal('transparent');
					expect(result.tunnel).to.be.an('object');
					expect(result.tunnel.ok).to.equal(true);
					expect(result.protocols).to.be.an('object');
					expect(result.protocols[proxyProtocol]).to.be.an('object');
					expect(result.protocols[proxyProtocol].ok).to.equal(true);
				} catch (error) {
					return done(error);
				}
				done();
			});
		});
	});

	describe('good proxy', function() {

		var proxyProtocols = ['http'];
		var proxyServer;

		before(function() {
			proxyServer = helpers.createProxyServer(5050, '127.0.0.2');
		});

		after(function() {
			proxyServer.close();
			proxyServer.http.close();
			proxyServer.https.close();
		});

		_.each(proxyProtocols, function(proxyProtocol) {

			it(proxyProtocol, function(done) {

				var proxy = {
					ipAddress: proxyServer[proxyProtocol].address().address,
					port: proxyServer[proxyProtocol].address().port,
					protocols: [proxyProtocol]
				};

				var options = {
					testUrl: 'https://127.0.0.1:3002/check',
					ipAddressCheckUrl: 'http://127.0.0.1:3001/check',
					requestOptions: {
						strictSSL: false,
						agentOptions: {
							rejectUnauthorized: false
						},
						timeout: 100
					}
				};

				ProxyVerifier.testAll(proxy, options, function(error, result) {

					try {
						expect(error).to.equal(null);
						expect(result).to.be.an('object');
						expect(result.anonymityLevel).to.equal('elite');
						expect(result.tunnel).to.be.an('object');
						expect(result.tunnel.ok).to.equal(true);
						expect(result.protocols).to.be.an('object');
						expect(result.protocols[proxyProtocol]).to.be.an('object');
						expect(result.protocols[proxyProtocol].ok).to.equal(true);
					} catch (error) {
						return done(error);
					}

					done();
				});
			});
		});
	});
});


test/ssl/index.js
'use strict';

var fs = require('fs');

module.exports = {
	cert: fs.readFileSync(__dirname + '/server.crt', 'utf8'),
	key: fs.readFileSync(__dirname + '/server.key', 'utf8')
};


httpStatusCodes.js
'use strict';

module.exports = {
	300: 'Multiple Choices',
	301: 'Moved Permanently',
	302: 'Found',
	303: 'See Other',
	304: 'Not Modified',
	305: 'Use Proxy',
	306: 'Switch Proxy',
	307: 'Temporary Redirect',
	308: 'Permanent Redirect',
	400: 'Bad Request',
	401: 'Unauthorized',
	402: 'Payment Required',
	403: 'Forbidden',
	404: 'Not Found',
	405: 'Method Not Allowed',
	406: 'Not Acceptable',
	407: 'Proxy Authentication Required',
	408: 'Request Timeout',
	409: 'Conflict',
	410: 'Gone may be used instead.',
	411: 'Length Required',
	412: 'Precondition Failed',
	413: 'Payload Too Large',
	414: 'URI Too Long',
	415: 'Unsupported Media Type',
	416: 'Range Not Satisfiable',
	417: 'Expectation Failed',
	418: 'I\'m a teapot',
	419: 'Authentication Timeout',
	421: 'Misdirected Request',
	422: 'Unprocessable Entity',
	423: 'Locked',
	424: 'Failed Dependency',
	426: 'Upgrade Required',
	428: 'Precondition Required',
	429: 'Too Many Requests',
	431: 'Request Header Fields Too Large',
	500: 'Internal Server Error',
	501: 'Not Implemented',
	502: 'Bad Gateway',
	503: 'Service Unavailable',
	504: 'Gateway Timeout',
	505: 'HTTP Version Not Supported',
	506: 'Variant Also Negotiates',
	507: 'Insufficient Storage',
	508: 'Loop Detected',
	510: 'Not Extended',
};


proxy-verifier.d.ts
import {Options as RequestOptions} from "request";

declare class ProxyVerifier {
    static testAll(proxy: ProxyVerifier.IProxy & ProxyVerifier.IProtocolsProxy, options: RequestOptions, cb: (error: any, result: ProxyVerifier.IAllResults) => void): void;
    static testAll(proxy: ProxyVerifier.IProxy & ProxyVerifier.IProtocolsProxy, cb: (error: any, result: ProxyVerifier.IAllResults) => void): void;

    static testProtocol(proxy: ProxyVerifier.IProtocolsProxy, options: RequestOptions, cb: (error: any, result: ProxyVerifier.Result) => void): void;
    static testProtocol(proxy: ProxyVerifier.IProtocolsProxy, cb: (error: any, result: ProxyVerifier.Result) => void): void;

    static testProtocols(proxy: ProxyVerifier.IProxy, options: RequestOptions, cb: (error: any, result: ProxyVerifier.IProtocolResult) => void): void;
    static testProtocols(proxy: ProxyVerifier.IProxy, cb: (error: any, result: ProxyVerifier.IProtocolResult) => void): void;

    static testAnonymityLevel(proxy: ProxyVerifier.IProxy, options: RequestOptions, cb: (error: any, result: string) => void): void;
    static testAnonymityLevel(proxy: ProxyVerifier.IProxy, cb: (error: any, result: string) => void): void;

    static testTunnel(proxy: ProxyVerifier.IProxy, options: RequestOptions, cb: (error: any, result: ProxyVerifier.Result) => void): void;
    static testTunnel(proxy: ProxyVerifier.IProxy, cb: (error: any, result: ProxyVerifier.Result) => void): void;

    static test(proxy: ProxyVerifier.IProxy, options: ProxyVerifier.ITestOptions, cb: (error: any, result: ProxyVerifier.ICustomTestResult) => void): void;
    static test(proxy: ProxyVerifier.IProxy, cb: (error: any, result: ProxyVerifier.ICustomTestResult) => void): void;
}

export = ProxyVerifier;

declare namespace ProxyVerifier {
    interface IBaseProxy {
        ipAddress: string;
        port: number;
        /**
         * Proxy-Authorization header
         */
        auth?: string
    }

    export interface IProxy extends IBaseProxy {
        protocol?: Protocol;
    }

    export interface IProtocolsProxy extends IBaseProxy {
        protocols?: Protocol[];
    }

    export type Protocol = "http" | "https" | "socks5" | "socks4";

    export type AnonymityLevel = "transparent" | "anonymous" | "elite";


    export interface IAllResults {
        anonymityLevel?: AnonymityLevel;
        protocols?: IProtocolResult;
        tunnel?: Result
    }

    export type Result = IWorkingResult | INotWorkingResult;

    interface IWorkingResult {
        ok: true
    }

    interface INotWorkingResult {
        ok: false;
        error: {
            message: string;
            code: string;
        }
    }

    export interface IProtocolResult {
        [key: string]: Result;
    }

    export interface ITestOptions {
        testUrl: string;
        testFn: (data: string, status: number, headers: IHeaders) => void;
    }

    interface IHeaders {
        [key: string]: string
    }

    interface ICustomTestBaseResult {
        data: string;
        status: number;
        headers: IHeaders;
    }

    export type ICustomTestResult = ICustomTestWorkingResult | ICustomTestNotWorkingResult;

    interface ICustomTestWorkingResult extends ICustomTestBaseResult {
        ok: true
    }

    interface ICustomTestNotWorkingResult extends ICustomTestBaseResult {
        ok: false;
        error: {
            message: string;
            code: string;
        }
    }

}


scripts/update-data.js
'use strict';

var _ = require('underscore');
var Address4 = require('ip-address').Address4;
var Address6 = require('ip-address').Address6;
var async = require('async');
var colors = require('colors');
var fs = require('fs');
var https = require('https');
var lazy = require('lazy');
var path = require('path');
var program = require('commander');
var url = require('url');
var yauzl = require('yauzl');

var utils = require('../lib/utils');

var packageInfo = JSON.parse(fs.readFileSync(__dirname + '/../package.json', 'utf8'));

console.log(__dirname);

program
	.version(packageInfo.version)
	.option('--debug', 'debugging mode')
	.parse(process.argv);

var dataDir = __dirname + '/../data';
var tmpDir = dataDir + '/tmp';

if (!fs.existsSync(dataDir)) {
	fs.mkdirSync(dataDir);
}

if (!fs.existsSync(tmpDir)) {
	fs.mkdirSync(tmpDir);
}

function download(cb) {

	console.log('Downloading...');

	var downloadUrl = 'https://geolite.maxmind.com/download/geoip/database/GeoLite2-Country-CSV.zip';
	var fileName = path.basename(url.parse(downloadUrl).pathname);
	var file = path.join(tmpDir, fileName);

	fs.stat(file, function(error, stat) {

		if (!error && program.debug) {
			// File already exists.
			return cb(null, file);
		}

		var req = https.get(downloadUrl, function(res) {

			res.pipe(fs.createWriteStream(file))
				.on('error', cb)
				.on('close', function() {
					cb(null, file);
				});
		});

		req.on('error', cb);
	});
}

function extract(zipArchive, cb) {

	console.log('Extracting...');

	// Unzip and grab the files that we need.

	var fileNames = [
		'GeoLite2-Country-Blocks-IPv4.csv',
		'GeoLite2-Country-Blocks-IPv6.csv',
		'GeoLite2-Country-Locations-en.csv',
		'COPYRIGHT.txt',
		'LICENSE.txt'
	];

	var extracted = {};
	var called = false;

	function done(error, fileName) {

		if (called) {
			// Do nothing, because we've already called cb().
			return;
		}

		if (error) {
			called = true;
			return cb(error);
		}

		extracted[fileName] = true;

		var allFilesExtracted = _.every(fileNames, function(fileName) {
			return extracted[fileName];
		});

		if (allFilesExtracted) {
			called = true;
			return cb();
		}
	};

	yauzl.open(zipArchive, function(error, openedZipArchive) {

		if (error) {
			return done(error);
		}

		openedZipArchive.on('entry', function(entry) {

			var fileName = path.basename(entry.fileName);
			var isDirectory = /\/$/.test(fileName);

			if (!isDirectory && _.contains(fileNames, fileName)) {

				openedZipArchive.openReadStream(entry, function(error, readStream) {

					if (error) {
						return done(error);
					}

					var file = path.join(tmpDir, fileName);

					fs.stat(file, function(error, stat) {

						if (!error && program.debug) {
							// File already exists.
							return done(null, fileName);
						}

						readStream.pipe(fs.createWriteStream(file))
							.on('error', cb)
							.on('close', function() {
								done(null, fileName);
							});
					});
				});
			}
		});
	});
}

function processData(cb) {

	console.log('Processing data...');

	async.parallel({
		countryBlocksIpv4: processCountryBlocksIpv4,
		countryBlocksIpv6: processCountryBlocksIpv6,
		countryLocations: processCountryLocations
	}, cb);
}

function processCountryBlocksIpv4(cb) {

	var file = path.join(tmpDir, 'GeoLite2-Country-Blocks-IPv4.csv');
	var countryBlocksIpv4 = [];

	lazy(fs.createReadStream(file))
		.lines
		.map(function(byteArray) {
			return (new Buffer(byteArray)).toString('utf8');
		})
		.skip(1)
		.map(function(line) {

			var data = line.split(',');

			if (!data || data.length < 2) {
				console.log('Bad line '.red, line);
				return;
			}

			var geonameId = parseInt(data[1].replace(/"/g, ''));
			var ipv4Cidr = data[0].replace(/"/g, '');
			var ipv4Address = new Address4(ipv4Cidr);
			var ipv4StartAddress = ipv4Address.startAddress().correctForm();
			var ipv4EndAddress = ipv4Address.endAddress().correctForm();

			var block = {
				geoname_id: geonameId,
				ipv4_start_int: utils.ipv4ToInt(ipv4StartAddress)
			};

			if (ipv4StartAddress !== ipv4EndAddress) {

				block.ipv4_end_int = utils.ipv4ToInt(ipv4EndAddress);
			}

			countryBlocksIpv4.push(block);
		})
		.on('pipe', function() {

			cb(null, countryBlocksIpv4);
		});
}

function processCountryBlocksIpv6(cb) {

	var file = path.join(tmpDir, 'GeoLite2-Country-Blocks-IPv6.csv');
	var countryBlocksIpv6 = [];

	lazy(fs.createReadStream(file))
		.lines
		.map(function(byteArray) {
			return (new Buffer(byteArray)).toString('utf8');
		})
		.skip(1)
		.map(function(line) {

			var data = line.split(',');

			if (!data || data.length < 2) {
				console.log('Bad line '.red, line);
				return;
			}

			var geonameId = parseInt(data[1].replace(/"/g, ''));
			var ipv6Cidr = data[0].replace(/"/g, '');
			var ipv6Address = new Address6(ipv6Cidr);
			var ipv6StartAddress = ipv6Address.startAddress().correctForm();
			var ipv6EndAddress = ipv6Address.endAddress().correctForm();

			var block = {
				geoname_id: geonameId,
				ipv6_start_int_arr: utils.ipv6ToIntArray(ipv6StartAddress)
			};

			if (ipv6StartAddress !== ipv6EndAddress) {

				block.ipv6_end_int_arr = utils.ipv6ToIntArray(ipv6EndAddress);
			}

			countryBlocksIpv6.push(block);
		})
		.on('pipe', function() {

			cb(null, countryBlocksIpv6);
		});
}

function processCountryLocations(cb) {

	var file = path.join(tmpDir, 'GeoLite2-Country-Locations-en.csv');
	var countryLocations = {};

	lazy(fs.createReadStream(file))
		.lines
		.map(function(byteArray) {
			return (new Buffer(byteArray)).toString('utf8');
		})
		.skip(1)
		.map(function(line) {

			var data = line.split(',');

			if (!data || data.length < 2) {
				console.log('Bad line '.red, line);
				return;
			}

			var geonameId = parseInt(data[0].replace(/"/g, ''));
			var isoCode = data[4].replace(/"/g, '').toLowerCase();

			countryLocations[geonameId] = isoCode;
		})
		.on('pipe', function() {

			cb(null, countryLocations);
		});
}

function buildDataFiles(data, cb) {

	console.log('Building data files...');

	async.parallel([
		_.bind(buildIpv4DataFile, undefined, data),
		_.bind(buildIpv6DataFile, undefined, data)
	], function(error) {

		if (error) {
			return cb(error);
		}

		cb();
	});
}

function buildIpv4DataFile(data, cb) {

	var json = _.map(data.countryBlocksIpv4, function(item) {

		var block = {
			c: data.countryLocations[item.geoname_id],
			s: item.ipv4_start_int
		};

		if (item.ipv4_end_int) {
			block.e = item.ipv4_end_int;
		}

		return block;
	});

	var file = path.join(dataDir, 'country-ipv4.json');

	fs.writeFile(file, JSON.stringify(json), 'utf8', cb);
}

function buildIpv6DataFile(data, cb) {

	var json = _.map(data.countryBlocksIpv6, function(item) {

		var block = {
			c: data.countryLocations[item.geoname_id],
			s: item.ipv6_start_int_arr
		};

		if (item.ipv6_end_int_arr) {
			block.e = item.ipv6_end_int_arr;
		}

		return block;
	});

	var file = path.join(dataDir, 'country-ipv6.json');

	fs.writeFile(file, JSON.stringify(json), 'utf8', cb);
}

function moveTmpFilesToDataDir(cb) {

	console.log('Moving files to data directory...');

	var tmpFileNames = [
		'COPYRIGHT.txt',
		'LICENSE.txt'
	];

	async.each(tmpFileNames, function(tmpFileName, next) {

		var tmpFile = path.join(tmpDir, tmpFileName);
		var file = path.join(dataDir, tmpFileName);

		fs.rename(tmpFile, file, next);

	}, function(error) {

		if (error) {
			return cb(error);
		}

		cb();
	});
}

function cleanup() {

	console.log('Cleaning up...');

	if (!program.debug) {

		var tmpFiles = fs.readdirSync(tmpDir);

		_.each(tmpFiles, function(tmpFile) {
			fs.unlinkSync(path.join(tmpDir, tmpFile));
		});

		fs.rmdirSync(tmpDir);
	}
}

async.seq(download, extract, processData, buildDataFiles, moveTmpFilesToDataDir)(function(error) {

	cleanup();

	if (error) {
		console.error(error.message.red);
		return process.exit(1);
	}

	console.log('Done'.green);
	process.exit(0);
});


index.js
'use strict';

var _ = require('underscore');
var async = require('async');
var deprecate = require('depd')('ProxyVerifier');
var geoip = require('geoip-lite');
var ProxyAgent = require('proxy-agent');
var request = require('request');
var url = require('url');

var httpStatusCodes = require('./httpStatusCodes');

var ProxyVerifier = module.exports = {

	_defaultTestUrl: 'https://ifconfig.me/all.json',
	_ipAddressCheckUrl: 'https://ifconfig.me/all.json',
	_tunnelTestUrl: 'https://ifconfig.me/all.json',

	/*
		Array of header keys for exact matching.
	*/
	_proxyHeaders: ['via'],

	/*
		Array of header keywords for loose matching.
	*/
	_proxyRelatedHeaderKeywords: ['proxy'],

	testAll: function(proxy, options, cb) {

		if (_.isUndefined(cb)) {
			cb = options;
			options = null
		}

		options || (options = {});

		// Tells the request module to measure the elapsed time for each request.
		options.time = true;

		proxy = ProxyVerifier.normalizeProxy(proxy);

		var testProtocolsOptions = _.extend({}, options, { includeAll: true });

		ProxyVerifier.testProtocols(proxy, testProtocolsOptions, function(error, protocolsResult) {

			if (error) {
				return cb(error);
			}

			var workingProtocols = _.filter(_.keys(protocolsResult), function(protocol) {
				return protocolsResult[protocol].ok === true;
			});

			proxy.protocols = workingProtocols;

			var asyncTests = {};

			if (!_.isEmpty(workingProtocols)) {

				asyncTests.anonymityLevel = function(next) {
					var testAnonymityLevelOptions = _.extend({}, options);
					ProxyVerifier.testAnonymityLevel(proxy, testAnonymityLevelOptions, next);
				};

				asyncTests.tunnel = function(next) {
					ProxyVerifier.testTunnel(proxy, options, next);
				};
			}

			async.parallel(asyncTests, function(error, results) {

				if (error) {
					return cb(error);
				}

				results.protocols = _.object(_.map(_.keys(protocolsResult), function(protocol) {
					return [protocol, _.omit(protocolsResult[protocol], 'data', 'status', 'headers')]
				}));

				if (_.isEmpty(workingProtocols)) {
					results.anonymityLevel = null;
					results.tunnel = { ok: false };
				}

				cb(null, results);
			});
		});
	},

	testProtocols: function(proxy, options, cb) {

		if (_.isUndefined(cb)) {
			cb = options;
			options = null
		}

		options || (options = {});
		proxy = ProxyVerifier.normalizeProxy(proxy);

		if (!_.isArray(proxy.protocols)) {
			throw new Error('Invalid "protocols" attribute: Array expected.');
		}

		if (!proxy.protocols || _.isEmpty(proxy.protocols)) {
			throw new Error('Must specify some protocols to test.');
		}

		var tests = _.object(_.map(proxy.protocols, function(protocol) {
			var _proxy = _.extend({}, proxy, { protocol: protocol });
			return [protocol, _.bind(ProxyVerifier.testProtocol, undefined, _proxy, options)];
		}));

		async.parallel(tests, cb);
	},

	testProtocol: function(proxy, options, cb) {

		ProxyVerifier.test(proxy, options, cb);
	},

	testAnonymityLevel: function(proxy, options, cb) {

		if (_.isUndefined(cb)) {
			cb = options;
			options = null
		}

		options = (options || {});

		async.parallel({

			myIpAddress: function(next) {

				if (options.myIpAddress) {
					// Already have our IP address for comparison.
					return next(null, options.myIpAddress);
				}

				ProxyVerifier.getMyIpAddress(options, next);
			},

			test: function(next) {

				var testOptions = _.extend({}, options, { includeAll: true });

				ProxyVerifier.test(proxy, testOptions, function(error, result) {

					if (error || result.error) {

						if (!error) {

							error = new Error(result.error.message);

							if (result.error.code) {
								error.code = result.error.code;
							}
						}

						return next(error);
					}

					next(null, result);
				});
			}

		}, function(error, results) {

			if (error) {
				return cb(error);
			}

			var anonymityLevel;
			var myIpAddress = results.myIpAddress;
			var withProxy = results.test;
			var ipAddressMatches = !!myIpAddress && withProxy.data.ip_addr === myIpAddress;
			var ipAddressRegEx = /[:.0-9a-z%]+/g;
			var anyHeaderContainsIpAddress = !!_.chain(withProxy.data).find(function(value, key) {
				var matches = value.toString().match(ipAddressRegEx);
				return _.contains(matches, myIpAddress);
			}).value();

			// If the IP address matches ours or if any of the headers contain our IP address, then "transparent".
			if (ipAddressMatches || anyHeaderContainsIpAddress) {
				anonymityLevel = 'transparent';
			} else {

				var proxyHeaders = ProxyVerifier._proxyHeaders;
				var proxyKeywords = ProxyVerifier._proxyRelatedHeaderKeywords;
				var headerKeys = _.keys(withProxy.data);

				var hasProxyHeaders = _.some(proxyHeaders, function(proxyHeader) {
					return _.contains(headerKeys, proxyHeader) || _.some(headerKeys, function(headerKey) {
						return _.some(proxyKeywords, function(proxyKeyword) {
							return headerKey.indexOf(proxyKeyword) !== -1;
						});
					});
				});

				if (hasProxyHeaders) {
					anonymityLevel = 'anonymous';
				} else {
					anonymityLevel = 'elite';
				}
			}

			cb(null, anonymityLevel);
		});
	},

	getMyIpAddress: function(options, cb) {

		if (_.isUndefined(cb)) {
			cb = options;
			options = null
		}

		options = _.defaults(options || {}, {
			ipAddressCheckUrl: ProxyVerifier._ipAddressCheckUrl,
			ipAddressCheckFn: ProxyVerifier._getIpAddressFromCheckProxyServiceResponse
		});

		ProxyVerifier.request('get', options.ipAddressCheckUrl, options, function(error, data, status, headers) {

			if (error) {
				return cb(error);
			}

			try {
				var ipAddress = options.ipAddressCheckFn(data, status, headers);
			} catch (error) {
				return cb(error);
			}

			cb(null, ipAddress);
		});
	},

	testTunnel: function(proxy, options, cb) {

		if (_.isUndefined(cb)) {
			cb = options;
			options = null
		}

		options = _.defaults(options || {}, {
			testUrl: ProxyVerifier._tunnelTestUrl,
			maxAttempts: 5,
			waitTimeBetweenAttempts: 0
		});

		ProxyVerifier.test(proxy, options, cb);
	},

	test: function(proxy, options, cb) {

		if (_.isUndefined(cb)) {
			cb = options;
			options = null
		}

		if (!_.isObject(proxy)) {
			throw new Error('Expected "proxy" to be an object.');
		}

		if (!_.isObject(options)) {
			throw new Error('Expected "options" to be an object.');
		}

		if (!_.isObject(cb)) {
			throw new Error('Expected "cb" to be a function.');
		}

		options = _.defaults(options || {}, {
			testUrl: ProxyVerifier._defaultTestUrl,
			testFn: ProxyVerifier._checkProxyServiceResponse,
			maxAttempts: 1,
			waitTimeBetweenAttempts: 0
		});

		var requestOptions = _.extend({}, options.requestOptions || {}, {
			proxy: ProxyVerifier.normalizeProxy(proxy)
		});

		var result;
		var numAttempts = 0;

		async.until(function(next) {

			// Retry the test until successful or have reached the maximum number of attempts.
			var doContinue = (result && result.ok === true) || numAttempts >= options.maxAttempts;
			next(null, doContinue);

		}, function(nextAttempt) {

			numAttempts++;

			ProxyVerifier.request('get', options.testUrl, requestOptions, function(error, data, status, headers) {

				if (!error) {

					try {
						options.testFn(data, status, headers);
					} catch (testError) {
						error = testError;
					}
				}

				if (error) {

					result = {
						ok: false,
						error: {
							message: error.message,
							code: error.code
						}
					};

				} else {

					result = {
						ok: true
					};
				}

				if (options.includeAll) {
					result.data = data;
					result.status = status;
					result.headers = headers;
				}

				// Don't wait in the case of success.
				setTimeout(nextAttempt, result.ok === true ? 0 : options.waitTimeBetweenAttempts);
			});

		}, function(error) {

			if (error) {
				return cb(error);
			}

			cb(null, result);
		});
	},

	request: function(method, uri, options, cb) {

		cb = _.last(arguments);
		options || (options = {});

		if (options.proxy) {
			options.proxy = ProxyVerifier.normalizeProxy(options.proxy);
		}

		var requestOptions = _.extend({}, _.omit(options, 'proxy', 'data'), {
			method: method.toUpperCase(),
			url: uri,
			headers: {}
		});

		if (options.proxy) {

			var proxy = options.proxy;
			var proxyProtocol = _.first(proxy.protocols);
			var agentOptions = _.extend(
				{},
				url.parse(proxyProtocol + '://' + proxy.ipAddress + ':' + proxy.port),
				options.agentOptions || {}
			);

			requestOptions.agent = new ProxyAgent(agentOptions);

			if (proxy.auth) {
				requestOptions.headers['Proxy-Authorization'] = proxy.auth;
			}
		}

		if (options.data) {

			switch (method) {

				case 'GET':
				case 'DELETE':
					requestOptions.qs = options.data;
					break;

				case 'POST':
				case 'PUT':
					requestOptions.json = options.data;
					break;
			}
		}

		var req = request(requestOptions);
		var done = _.once(cb);

		req.on('response', function(res) {

			res.setEncoding('utf8');

			var responseData = '';
			var status = res.statusCode;
			var headers = res.headers;

			res.on('data', function(chunk) {
				responseData += chunk;
			});

			res.on('end', function() {

				res.destroy();

				if (headers['content-type'] && headers['content-type'].indexOf('application/json') !== -1) {

					try {
						responseData = JSON.parse(responseData);
					} catch (error) {
						return done(error);
					}
				}

				done(null, responseData, status, headers);
			});
		});

		req.on('error', done);
		req.on('timeout', req.abort);
		req.end();

		return req;
	},

	normalizeProxy: function(proxy) {

		proxy = ProxyVerifier._deepClone(proxy);

		return {
			ipAddress: proxy.ipAddress || proxy.ip_address || null,
			port: proxy.port || null,
			protocols: proxy.protocols || (proxy.protocol && [proxy.protocol]) || null,
			auth: proxy.auth || null
		};
	},

	_deepClone: function(object) {

		return JSON.parse(JSON.stringify(object));
	},

	_getIpAddressFromCheckProxyServiceResponse: function(data, status, headers) {

		ProxyVerifier._checkProxyServiceResponse(data, status, headers);
		return data.ip_addr;
	},

	_checkProxyServiceResponse: function(data, status, headers) {

		var error;

		if (status >= 300) {
			error = new Error(httpStatusCodes[status] || '');
			error.code = 'HTTP_ERROR_' + status;
		} else if (status !== 200 && !_.isObject(data) && !_.has(data, 'ip_addr')) {
			error = new Error('Failed to reach proxy check service.');
			error.code = 'FAILED_TO_REACH_PROXY_SERVICE';
		}

		if (!_.isUndefined(error)) {
			throw error;
		}
	}
};

// For backwards compatibility, but with deprecated warnings.

ProxyVerifier.all = deprecate.function(
	ProxyVerifier.testAll,
	'all() has been deprecated; use testAll() instead'
);

ProxyVerifier.protocol = deprecate.function(
	ProxyVerifier.testProtocol,
	'protocol() has been deprecated; use testProtocol() instead'
);

ProxyVerifier.protocols = deprecate.function(
	ProxyVerifier.testProtocols,
	'protocols() has been deprecated; use testProtocols() instead'
);

ProxyVerifier.anonymityLevel = deprecate.function(
	ProxyVerifier.testAnonymityLevel,
	'anonymityLevel() has been deprecated; use testAnonymityLevel() instead'
);

ProxyVerifier.tunnel = deprecate.function(
	ProxyVerifier.testTunnel,
	'tunnel() has been deprecated; use testTunnel() instead'
);

var lookupCountry = function(proxy) {
	proxy = ProxyVerifier.normalizeProxy(proxy);
	var geo = geoip.lookup(proxy.ipAddress);
	return geo.country.toLowerCase();
};

ProxyVerifier.lookupCountry = deprecate.function(
	lookupCountry,
	'lookupCountry() has been deprecated and will be removed in a future release'
);

ProxyVerifier.country = deprecate.function(
	lookupCountry,
	'country() has been deprecated and will be removed in a future release'
);

ProxyVerifier.loadCountryData = deprecate.function(
	function(options, cb) {
		// Do nothing.
		_.delay(cb);
	},
	'loadCountryData() has been deprecated and will be removed in a future release'
);

ProxyVerifier.loadCountryDataSync = deprecate.function(
	function(options) {
		// Do nothing.
	},
	'loadCountryDataSync() has been deprecated and will be removed in a future release'
);
